### ISSUE TITLE
Where are activity phases actually defined?

### ISSUE THREAD

**Issue by caprica** [Bungie-net/api#623]
I am very interested in the completion status of the various phases in raids so I've been digging around the latest data looking for more information.

I eventually found a "phaseHash" in Destiny.Definitions.DestinyActivityInsertionPointDefinition, but it's not at all clear to me what type of entity that phaseHash points to.

**Comment by vthornheart-bng** [comment]
Greetings!  Yes, unfortunately we don't have any human readable information in our data to map to those phases (at least, not yet) - so in the end, those phases are just identifiers for the parts/sections of the raid.  The thing that will hopefully be useful for you is that they are consistent: a given section of the raid will always have the same phase identifier, which means you could map it to your own UI (be it strings that you write to represent the sections, checkboxes, a progress bar etc...)

So there's a bit of manual wire-up to be done depending on how you want to represent or use them, but you can use the consistency of those identifiers to your advantage when making that wire-up at least (which is something we didn't have until this most recent release!)

**Comment by caprica** [comment]
Ah OK, thanks a lot.

**Comment by caprica** [comment]
I have a follow-up question on the same subject...

Here is a fragment from GetProfile showing the Leviathan Raid milestone, retrieved today:
```
    "3660836525": {
      "order": 9000,
      "endDate": "2018-09-04T17:00:01Z",
      "startDate": "2018-08-28T17:00:01Z",
      "activities": [
        {
          "phases": [
            {
              "phaseHash": 2188993306,
              "complete": false
            },
            {
              "phaseHash": 3847906370,
              "complete": false
            },
            {
              "phaseHash": 1431486395,
              "complete": false
            },
            {
              "phaseHash": 4231923662,
              "complete": true
            }
          ],
          "challenges": [],
          "activityHash": 2693136601
        },
        {
          "phases": [],
          "challenges": [],
          "activityHash": 287649202
        },
        {
          "phases": [
            {
              "phaseHash": 1431486395,
              "complete": false
            },
            {
              "phaseHash": 2188993306,
              "complete": false
            },
            {
              "phaseHash": 3847906370,
              "complete": false
            },
            {
              "phaseHash": 4231923662,
              "complete": true
            }
          ],
          "challenges": [],
          "activityHash": 2693136604
        },
        {
          "phases": [],
          "challenges": [],
          "activityHash": 1699948563
        },
        {
          "phases": [
            {
              "phaseHash": 1431486395,
              "complete": false
            },
            {
              "phaseHash": 2188993306,
              "complete": false
            },
            {
              "phaseHash": 3847906370,
              "complete": false
            },
            {
              "phaseHash": 4231923662,
              "complete": false
            }
          ],
          "challenges": [],
          "activityHash": 757116822
        }
      ],
      "milestoneHash": 3660836525
    }
```
There are three separate "phases" blocks, presumably for the different raid instances with different encounter orders, I'm not sure.

Anyway, for this profile the first three encounters of the raid were completed, stopping at Calus.

Yet this data shows in the three separate phases blocks:

 * false, false, false, true
 * false, false, false, true
 * false, false, false, false

Only one of those I expect instead of course to be:

 * true, true, true, false

And the rest should all be:

 * false, false, false, false

I suppose it is the same issue as #136, but I was hoping the recently announced changes around raid phases would have resolved this.

Is there any hope of being able to use this profile data to check raid phase completion status in the near future? Or should I be looking elsewhere to try and get that data?

**Comment by vthornheart-bng** [comment]
Okay, looking into this further is reminding me of several investigations that I was doing around the time that #136 was filed.  Other unexpected situations came up, and I ended up tabling this and never coming back to it.

Now that I'm looking at it, what we're seeing is:

* The data that I'm currently returning as completed phases is actually information used to determine what phase you will *start at* if you rejoin/restart the activity.

* There's a separate set of data that I have to create new manual associations with to know which specific phases you've actually beaten: though it could be partially implied by the info above, there are situations where people could circumvent that which the info above wouldn't convey (for instance, joining in a raid party where someone else had a different insertion point)

In short, there's more manual work to be done: work that I started to look at back in September, and then dropped, never came back to, and eventually forgot about entirely.

I need to do more investigation to see whether the Raid Lairs follow a similar pattern in content, as they likely have the same flaw.  But I will see what we can find out.

**Comment by caprica** [comment]
Is there, or will there, also be a way to distinguish prestige/normal raid phases?

At the moment these phases in the profile are structured by milestone, and if I remember correctly there is no separate milestone for Leviathan raid vs Leviathan Prestige raid.

Where I'm coming from with this is in the D1 version of my web app you could basically see which chests had not been looted from normal or heroic raids - so if you happened to jump in at a checkpoint, you could still go back and raid again to pick up the chests that you skipped. I think it would be pretty useful to have that same detail back for the D2 version of my web app.

**Comment by vthornheart-bng** [comment]
The phase fixes is slated to be released October 2nd, though I welcome any and all bugs you can give.  I'm giving what I perceive in the data to be phases, but if there's additional steps you feel ought to be there or steps that don't seem right, let me know and I can do some more digging.

The Last Wish raid's weekly completion status will also start being returned on that date in Milestones.

For the names, on that same release date you should start seeing the more specific information for each variant returned in the displayProperties.name property on the activity definition.  That should be helpful for discerning between them, though you can also use things like the activityLightLevel property as well if you want something more programmatically useful!
